src/coreclr/System.Private.CoreLib/src/System/Variant.cs

```diff
@@ -78,323 +23,342 @@ internal partial struct Variant
+                    int error = pOle.GetRawDataRef<int>();
+                    return error == HResults.DISP_E_PARAMNOTFOUND ? Missing.Value : error;
+                case VarEnum.VT_BYREF | VarEnum.VT_ERROR:
+                    int refError = *(int*)pOle.GetRawDataRef<IntPtr>();
+                    return refError == HResults.DISP_E_PARAMNOTFOUND ? Missing.Value : refError;
+
+                case VarEnum.VT_VOID:
+                case VarEnum.VT_BYREF | VarEnum.VT_VOID:
+                    return null; // CV_VOID
+
+                default:
+                    throw new ArgumentException(SR.ComVariant_UnsupportedType);
             }
         }
 
-        // Helper code for marshaling VARIANTS to managed objects (we use
-        // managed variants as an intermediate type.
-        internal static object? MarshalHelperConvertVariantToObject(ref Variant v)
-        {
-            return v.ToObject();
-        }
-
         // Helper code: on the back propagation path where a VT_BYREF VARIANT*
         // is marshaled to a "ref Object", we use this helper to force the
         // updated object back to the original type.
-        internal static void MarshalHelperCastVariant(object pValue, int vt, ref Variant v)
+        internal static void MarshalHelperCastVariant(object pValue, int vt, out ComVariant v)
         {
-            if (!(pValue is IConvertible iv))
+            Debug.Assert((VarEnum)vt != VarEnum.VT_VARIANT, "Should be handled at native side.");
+
+            if (pValue is not IConvertible iv)
             {
-                switch (vt)
+                switch ((VarEnum)vt)
                 {
-                    case 9: /*VT_DISPATCH*/
+                    case VarEnum.VT_DISPATCH:
                         Debug.Assert(OperatingSystem.IsWindows());
-                        v = new Variant(new DispatchWrapper(pValue));
-                        break;
-
-                    case 12: /*VT_VARIANT*/
-                        v = new Variant(pValue);
+                        v = ComVariant.CreateRaw(VarEnum.VT_DISPATCH,
+                            pValue is null ? IntPtr.Zero : Marshal.GetIDispatchForObject(pValue));
                         break;
 
-                    case 13: /*VT_UNKNOWN*/
-                        v = new Variant(new UnknownWrapper(pValue));
+                    case VarEnum.VT_UNKNOWN:
+                        v = ComVariant.CreateRaw(VarEnum.VT_UNKNOWN,
+                            pValue is null ? IntPtr.Zero : Marshal.GetIUnknownForObject(pValue));
                         break;
 
-                    case 36: /*VT_RECORD*/
-                        v = new Variant(pValue);
+                    case VarEnum.VT_RECORD:
+                        MarshalHelperConvertObjectToVariant(pValue, out v);
+                        if (v.VarType != VarEnum.VT_RECORD)
+                            throw new InvalidCastException(SR.InvalidCast_CannotCoerceByRefVariant);
                         break;
```

Comment:
I'm trying to restore the behavior here, and relying on the location to be cleaned up by RAII at native side.
Is there any better option for this?

Reply:
What is the previous behavior in this case? I'm not following the issue being hit here.

Reply:
I don't see any destructors in MarshalOleRefVariantForObject. Is that where you are talking about?

Reply:
Previously in `MarshalOleRefVariantForObject`, if `SafeVariantChangeType` fails, would the data (potentially BSTR or something else) in `vtmp` be leaked? `SafeVariantClear` won't be called in throwing path.

I've concluded that `SafeVariantChangeType` is unnecessary and will always fail. The content of the VARIANT should be cleared at managed side in failing path.